urlfd018a75f7_highlanderkev_StockExVirtuoso_tgz-pJ8-driver_Phase4J8/src/client/UserSim.java

```diff
@@ -0,0 +1,229 @@
+        orderCxlCount++;
+    }
+
+    private void makeOrder() throws Exception {
+        ArrayList<String> list = user.getProductList();
+        String product = list.get((int) (Math.random() * list.size()));
+        BookSide side = makeRandomSide();
+        Price p = makeRandomOrderPrice(side, product);
+        int v = makeRandomVolume(product);
+
+        user.submitOrder(product, p, v, side);
+        orderCount++;
+    }
+
+    private void makeQuoteCancel() throws Exception {
+        ArrayList<String> list = user.getProductList();
+        String product = list.get((int) (Math.random() * list.size()));
+        user.submitQuoteCancel(product);
+        quoteCxlCount++;
+    }
+
+    private void makeQuote() throws Exception {
+        ArrayList<String> list = user.getProductList();
+        String product = list.get((int) (Math.random() * list.size()));
+        Price bp = makeRandomPrice(BookSide.BUY, product);
+        int bv = makeRandomVolume(product);
+
+        Price sp = makeRandomPrice(BookSide.SELL, product);
+        int sv = makeRandomVolume(product);
+
+
+
+        try {
+            if (bp.greaterOrEqual(sp)) {
+                bp = sp.subtract(PriceFactory.makeLimitPrice("0.01"));
+            }
+            user.submitQuote(product, bp, bv, sp, sv);
+            quoteCount++;
+        } catch (Exception ex) {
+            Logger.getLogger(UserSim.class.getName()).log(Level.SEVERE, null, ex);
+        }
+    }
+
+    private BookSide makeRandomSide() {
+        if (Math.random() < 0.5) {
+            return BookSide.BUY;
+        } else {
+            return BookSide.SELL;
+        }
+    }
+
+    private Price makeRandomOrderPrice(BookSide side, String product) throws Exception {
+
+        if (Math.random() < 0.1) {
+            return PriceFactory.makeMarketPrice();
+        } else {
+            return makeRandomPrice(side, product);
+        }
+    }
+
+    private Price makeRandomPrice(BookSide side, String product) throws Exception {
+
+        double priceBase = (side == BookSide.BUY ? UserSimSettings.getBuyPriceBase(product) : UserSimSettings.getSellPriceBase(product));
+
+        double price = priceBase * (1 - UserSimSettings.priceVariance);
+        price += priceBase * (UserSimSettings.priceVariance * 2) * Math.random();
+
+        return PriceFactory.makeLimitPrice(String.format("%.2f", price));
+    }
+
+    private int makeRandomVolume(String product) {
+        int vol = (int) (UserSimSettings.getVolumeBase(product) * (1 - UserSimSettings.volumeVariance));
+        vol += UserSimSettings.getVolumeBase(product) * (UserSimSettings.volumeVariance * 2) * Math.random();
+        return vol;
+    }
+
+    private void waitRandomTime() {
+        int waitTime = (int) ((0.75 * waitBase) + (0.5 * waitBase * Math.random()));
+
+        synchronized (this) {
+            try {
+                wait(waitTime);
+            } catch (InterruptedException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)